magma/lib/magma/grammar.rb

```diff
@@ -1,5 +1,105 @@
+
 class Magma
   class Grammar < Sequel::Model
+    class Rule
+      def self.to_schema
+        {
+          gnomon_rules: {
+            type: "object",
+            patternProperties: {
+              "^\\w+$": {
+                type: "string"
+              }
+            }
+          }
+        }
+      end
+    end
+    class Token
+      def self.to_schema
+        {
+          gnomon_tokens: {
+            type: "object",
+            patternProperties: {
+              "^[A-Z]+$": {
+                "$ref": "#/definitions/gnomon_token"
+              }
+            }
+          },
+          gnomon_token: {
+            type: "object",
+            properties: {
+              label: { type: "string" },
+              values: {
+                type: "object",
+                patternProperties: {
+                  ".*": { type: "string" }
+                }
+              }
+            }
+          }
+        }
+      end
+    end
+    def self.to_schema
+      {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+	title: "Gnomon Grammar",
+        description: "This grammar defines a set of tokens and rules which can be used to match or create valid identifiers for a Mount Etna project.",
+        definitions: [
+          Magma::Grammar::Rule,
+          Magma::Grammar::Token,
+        ].map(&:to_schema).reduce(&:merge),
```

Comment:
Hmm, perhaps. I at some point decided it was redundant since it could be inferred from tokens with identical labels, but that seems brittle, might be better to be explicit about it.